### PR TITLE
APM-668 Search memory accounting

### PR DESCRIPTION
**Pull request by MBkkt**
### Scope & Purpose

Check ticket

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.11: *(Please link PR)*
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR: https://github.com/arangodb/enterprise/pull/1314
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: Documentation/Metrics/arangodb_search_consolidations_memory.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+name: arangodb_search_consolidations_memory
+introducedIn: "3.12.0"
+help: |
+  Count of memory in bytes that used for consolidate this index.
```
Comment (Dronplane) [inline Documentation/Metrics/arangodb_search_consolidations_memory.yaml:4]:
```suggestion
  Count of memory in bytes that is used for consolidating this index.
```

File: Documentation/Metrics/arangodb_search_consolidations_memory.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+unit: number
+type: gauge
+category: ArangoSearch
+complexity: advanced
+exposedBy:
+  - dbserver
+  - single
+description: |
+  Count of memory in bytes that used for consolidate this index.
```
Comment (Dronplane) [inline Documentation/Metrics/arangodb_search_consolidations_memory.yaml:13]:
```suggestion
  Count of memory in bytes that is used for consolidating this index.
```

File: Documentation/Metrics/arangodb_search_writers_memory.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  Count of memory in bytes that used for write in this index.
+unit: number
+type: gauge
+category: ArangoSearch
+complexity: advanced
+exposedBy:
+  - dbserver
+  - single
+description: |
+  Count of memory in bytes that used for write in this index.
```
Comment (Dronplane) [inline Documentation/Metrics/arangodb_search_writers_memory.yaml:13]:
```suggestion
  Count of memory in bytes that is used for writing to this index.
```

File: Documentation/Metrics/arangodb_search_readers_memory.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+name: arangodb_search_writers_memory
```
Comment (Dronplane) [inline Documentation/Metrics/arangodb_search_readers_memory.yaml:1]:
```suggestion
name: arangodb_search_readers_memory
```

File: Documentation/Metrics/arangodb_search_readers_memory.yaml [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+introducedIn: "3.12.0"
+help: |
+  Count of memory in bytes that used for read from this index.
```
Comment (Dronplane) [inline Documentation/Metrics/arangodb_search_readers_memory.yaml:4]:
```suggestion
  Count of memory in bytes that is used for reading from this index.
```

File: Documentation/Metrics/arangodb_search_readers_memory.yaml [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+unit: number
+type: gauge
+category: ArangoSearch
+complexity: advanced
+exposedBy:
+  - dbserver
+  - single
+description: |
+  Count of memory in bytes that used for read from this index.
```
Comment (Dronplane) [inline Documentation/Metrics/arangodb_search_readers_memory.yaml:13]:
```suggestion
  Count of memory in bytes that is used for reading from this index.
```

File: Documentation/Metrics/arangodb_search_writers_memory.yaml [hunk 7]
```diff
@@ -0,0 +1,13 @@
+name: arangodb_search_writers_memory
+introducedIn: "3.12.0"
+help: |
+  Count of memory in bytes that used for write in this index.
```
Comment (Dronplane) [inline Documentation/Metrics/arangodb_search_writers_memory.yaml:4]:
```suggestion
  Count of memory in bytes that is used for writing to this index.
```

**Review by Dronplane — CHANGES_REQUESTED** [review]

File: arangod/IResearch/IResearchFeature.h [hunk 8]
```diff
@@ -180,6 +180,14 @@ class IResearchFeature final : public ArangodFeature {
   // found during recovery.
   std::vector<std::string> _skipRecoveryItems;
 
+  // number of links/indexes currently out of sync
+  metrics::Gauge<uint64_t>& _outOfSyncLinks;
+
+#ifdef USE_ENTERPRISE
+  irs::IResourceManager& _columnsCacheMemoryUsed;
```
Comment (Dronplane) [inline arangod/IResearch/IResearchFeature.h:187]: Should this reference be initialized somehow?

File: arangod/IResearch/IResearchMetricStats.h [hunk 9]
```diff
@@ -149,4 +160,5 @@ DECLARE_GAUGE(arangodb_search_consolidation_time, uint64_t,
 
 inline constexpr std::string_view kSearchStats = "arangodb_search_link_stats";
 
+
```
Comment (Dronplane) [inline arangod/IResearch/IResearchMetricStats.h:163]: accidental edit?

File: arangod/IResearch/ResourceManager.hpp [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    return true;
+  }
+
+  void Decrease(size_t v) noexcept final {
+    [[maybe_unused]] auto const was = fetch_sub(v);
+    TRI_ASSERT(v <= was);
+  }
+};
+
+struct LimittedResourceManager final : metrics::Gauge<uint64_t>,
```
Comment (Dronplane) [inline arangod/IResearch/ResourceManager.hpp:44]:
```suggestion
struct LimitedResourceManager final : metrics::Gauge<uint64_t>,
```
Double `t` in the name looks like a typo

**Review by Dronplane — COMMENTED** [review]
One test case is missing

File: tests/js/client/shell/api/arangosearch-memory-metrics-noncluster.js [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+      assertTrue(value > 0);
+    },
+
+    tearDownAll: function () {
+      db._dropView(view);
+      db._drop(collection);
+    },
+
+    testSimple: function () {
+      let c = db._collection(collection);
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply tests/js/client/shell/api/arangosearch-memory-metrics-noncluster.js:66]: Needs to fix API in iresearch.
It's not part of APM ticket, so I suggest to fix it separate PR.

[the remark this answers is not in the document]
Comment (MBkkt) [reply tests/js/client/shell/api/arangosearch-memory-metrics-noncluster.js:65]: Done

**Pull request merged**